In [2]:
#Imports

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

## Data Exploratory Analysis

In [3]:
df=pd.read_csv("C:/Users/jwsam/Desktop/Ecom/kaggle_datasets/olist_order_reviews_dataset.csv")

In [4]:
df_order_reviews=pd.read_csv("C:/Users/jwsam/Desktop/Ecom/kaggle_datasets/olist_order_reviews_dataset.csv")

In [5]:
df_order_reviews.head(70)

review_id                          order_id  \
0   7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1   80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2   228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3   e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4   f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   
..                               ...                               ...   
65  9924ac6169f2edc95ca3394e529f9580  a3ba7fa5266d62188949da3d6ccc3884   
66  e12a8d938f3092e010627d001f91d137  b3dca2e199b13ddabab4ed77d79ee9f6   
67  109b5ce2dd11bb8460eff3b86da6fefc  25362fbf6aac4b01a28dee1e076acc26   
68  6d06808638ec0701bccd70bc8d462c28  97d2f8fe76f2f253b8291e17b5383884   
69  750d4311e8ddb6cba549ee86f06f0cca  4b32a4cdfc1b80066a8ee3a4b7a2094f   

    review_score review_comment_title  \
0              4                  NaN   
1              5                  NaN   
2              5                  NaN   
3              5                  NaN   
4              5                  NaN   
..           ...                  ...   
65             5                  NaN   
66             5                  NaN   
67             5      Ótimo Produto!    
68             1                  NaN   
69             5                  NaN   

                               review_comment_message review_creation_date  \
0                                                 NaN  2018-01-18 00:00:00   
1                                                 NaN  2018-03-10 00:00:00   
2                                                 NaN  2018-02-17 00:00:00   
3               Recebi bem antes do prazo estipulado.  2017-04-21 00:00:00   
4   Parabéns lojas lannister adorei comprar pela I...  2018-03-01 00:00:00   
..                                                ...                  ...   
65                                                NaN  2018-02-27 00:00:00   
66                                                NaN  2018-06-03 00:00:00   
67                           Recomendo o vendedor...   2018-08-17 00:00:00   
68  O produto não chegou no prazo estipulado e cau...  2016-12-09 00:00:00   
69  Produto entregue como solicitado, e com muita ...  2017-08-09 00:00:00   

   review_answer_timestamp  
0      2018-01-18 21:46:59  
1      2018-03-11 03:05:13  
2      2018-02-18 14:36:24  
3      2017-04-21 22:02:06  
4      2018-03-02 10:26:53  
..                     ...  
65     2018-03-02 08:33:35  
66     2018-06-06 12:41:24  
67     2018-08-17 21:47:08  
68     2016-12-19 14:10:43  
69     2017-08-09 12:03:45  

[70 rows x 7 columns]

In [6]:
df_order_reviews.info()

<class 'pandas.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype
---  ------                   --------------  -----
 0   review_id                99224 non-null  str  
 1   order_id                 99224 non-null  str  
 2   review_score             99224 non-null  int64
 3   review_comment_title     11568 non-null  str  
 4   review_comment_message   40977 non-null  str  
 5   review_creation_date     99224 non-null  str  
 6   review_answer_timestamp  99224 non-null  str  
dtypes: int64(1), str(6)
memory usage: 5.3 MB


For the review_id, order_id, review_score, review_creation_date, review_answer_timestamp the data is nont null with equal value of 99223 instances
While the review_comment_title and the review_comment_message have null values

In [7]:
df_order_reviews.describe()

review_score
count  99224.000000
mean       4.086421
std        1.347579
min        1.000000
25%        4.000000
50%        5.000000
75%        5.000000
max        5.000000

In [8]:
df_order_reviews['review_comment_title'].isna().sum()

np.int64(87656)

In [9]:
unique_values = df['review_comment_title'].dropna().unique()

In [10]:
unique_values.unique()

<StringArray>
[                'recomendo',           'Super recomendo',
   'Não chegou meu produto ',                     'Ótimo',
                'Muito bom.',                       'Bom',
              'Maravilhoso!',          'Entrega perfeita',
                  'MT lindo',           'Ótimo Produto! ',
 ...
 'Não posso recomendar aind',      'Rápido porém não bom',
           'entregue rapido',                        '7 ',
      'Entrega super Rápida',              'Nota máxima!',
        'muito bom produto ', 'Não foi entregue o pedido',
           'OTIMA EMBALAGEM',            'Foto enganosa ']
Length: 4527, dtype: str

In [11]:
df_order_reviews['review_comment_title'] = df_order_reviews['review_comment_title'].fillna("no title")

In [12]:
df_order_reviews

review_id                          order_id  \
0      7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1      80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2      228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3      e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4      f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   
...                                 ...                               ...   
99219  574ed12dd733e5fa530cfd4bbf39d7c9  2a8c23fee101d4d5662fa670396eb8da   
99220  f3897127253a9592a73be9bdfdf4ed7a  22ec9f0669f784db00fa86d035cf8602   
99221  b3de70c89b1510c4cd3d0649fd302472  55d4004744368f5571d1f590031933e4   
99222  1adeb9d84d72fe4e337617733eb85149  7725825d039fc1f0ceb7635e3f7d9206   
99223  efe49f1d6f951dd88b51e6ccd4cc548f  90531360ecb1eec2a1fbb265a0db0508   

       review_score review_comment_title  \
0                 4             no title   
1                 5             no title   
2                 5             no title   
3                 5             no title   
4                 5             no title   
...             ...                  ...   
99219             5             no title   
99220             5             no title   
99221             5             no title   
99222             4             no title   
99223             1             no title   

                                  review_comment_message review_creation_date  \
0                                                    NaN  2018-01-18 00:00:00   
1                                                    NaN  2018-03-10 00:00:00   
2                                                    NaN  2018-02-17 00:00:00   
3                  Recebi bem antes do prazo estipulado.  2017-04-21 00:00:00   
4      Parabéns lojas lannister adorei comprar pela I...  2018-03-01 00:00:00   
...                                                  ...                  ...   
99219                                                NaN  2018-07-07 00:00:00   
99220                                                NaN  2017-12-09 00:00:00   
99221  Excelente mochila, entrega super rápida. Super...  2018-03-22 00:00:00   
99222                                                NaN  2018-07-01 00:00:00   
99223  meu produto chegou e ja tenho que devolver, po...  2017-07-03 00:00:00   

      review_answer_timestamp  
0         2018-01-18 21:46:59  
1         2018-03-11 03:05:13  
2         2018-02-18 14:36:24  
3         2017-04-21 22:02:06  
4         2018-03-02 10:26:53  
...                       ...  
99219     2018-07-14 17:18:30  
99220     2017-12-11 20:06:42  
99221     2018-03-23 09:10:43  
99222     2018-07-02 12:59:13  
99223     2017-07-03 21:01:49  

[99224 rows x 7 columns]

In [13]:
df_order_reviews['review_comment_message'] = df_order_reviews['review_comment_message'].fillna("no message")

In [14]:
df_order_reviews

review_id                          order_id  \
0      7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1      80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2      228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3      e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4      f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   
...                                 ...                               ...   
99219  574ed12dd733e5fa530cfd4bbf39d7c9  2a8c23fee101d4d5662fa670396eb8da   
99220  f3897127253a9592a73be9bdfdf4ed7a  22ec9f0669f784db00fa86d035cf8602   
99221  b3de70c89b1510c4cd3d0649fd302472  55d4004744368f5571d1f590031933e4   
99222  1adeb9d84d72fe4e337617733eb85149  7725825d039fc1f0ceb7635e3f7d9206   
99223  efe49f1d6f951dd88b51e6ccd4cc548f  90531360ecb1eec2a1fbb265a0db0508   

       review_score review_comment_title  \
0                 4             no title   
1                 5             no title   
2                 5             no title   
3                 5             no title   
4                 5             no title   
...             ...                  ...   
99219             5             no title   
99220             5             no title   
99221             5             no title   
99222             4             no title   
99223             1             no title   

                                  review_comment_message review_creation_date  \
0                                             no message  2018-01-18 00:00:00   
1                                             no message  2018-03-10 00:00:00   
2                                             no message  2018-02-17 00:00:00   
3                  Recebi bem antes do prazo estipulado.  2017-04-21 00:00:00   
4      Parabéns lojas lannister adorei comprar pela I...  2018-03-01 00:00:00   
...                                                  ...                  ...   
99219                                         no message  2018-07-07 00:00:00   
99220                                         no message  2017-12-09 00:00:00   
99221  Excelente mochila, entrega super rápida. Super...  2018-03-22 00:00:00   
99222                                         no message  2018-07-01 00:00:00   
99223  meu produto chegou e ja tenho que devolver, po...  2017-07-03 00:00:00   

      review_answer_timestamp  
0         2018-01-18 21:46:59  
1         2018-03-11 03:05:13  
2         2018-02-18 14:36:24  
3         2017-04-21 22:02:06  
4         2018-03-02 10:26:53  
...                       ...  
99219     2018-07-14 17:18:30  
99220     2017-12-11 20:06:42  
99221     2018-03-23 09:10:43  
99222     2018-07-02 12:59:13  
99223     2017-07-03 21:01:49  

[99224 rows x 7 columns]

### Due to the fact that the review creation date and review answer timestamp have no relevance to what the consumers felt about the shopping experience

In [15]:
df_order_reviews.drop(columns=['review_creation_date','review_answer_timestamp'],inplace=True)

In [16]:
df_order_reviews

review_id                          order_id  \
0      7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1      80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2      228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3      e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4      f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   
...                                 ...                               ...   
99219  574ed12dd733e5fa530cfd4bbf39d7c9  2a8c23fee101d4d5662fa670396eb8da   
99220  f3897127253a9592a73be9bdfdf4ed7a  22ec9f0669f784db00fa86d035cf8602   
99221  b3de70c89b1510c4cd3d0649fd302472  55d4004744368f5571d1f590031933e4   
99222  1adeb9d84d72fe4e337617733eb85149  7725825d039fc1f0ceb7635e3f7d9206   
99223  efe49f1d6f951dd88b51e6ccd4cc548f  90531360ecb1eec2a1fbb265a0db0508   

       review_score review_comment_title  \
0                 4             no title   
1                 5             no title   
2                 5             no title   
3                 5             no title   
4                 5             no title   
...             ...                  ...   
99219             5             no title   
99220             5             no title   
99221             5             no title   
99222             4             no title   
99223             1             no title   

                                  review_comment_message  
0                                             no message  
1                                             no message  
2                                             no message  
3                  Recebi bem antes do prazo estipulado.  
4      Parabéns lojas lannister adorei comprar pela I...  
...                                                  ...  
99219                                         no message  
99220                                         no message  
99221  Excelente mochila, entrega super rápida. Super...  
99222                                         no message  
99223  meu produto chegou e ja tenho que devolver, po...  

[99224 rows x 5 columns]

In [17]:
df_order_reviews['review_comment_title'].value_counts()

review_comment_title
no title                     87656
Recomendo                      423
recomendo                      345
Bom                            293
super recomendo                270
                             ...  
Nota máxima!                     1
muito bom produto                1
Não foi entregue o pedido        1
OTIMA EMBALAGEM                  1
Foto enganosa                    1
Name: count, Length: 4528, dtype: int64

In [18]:
df_order_reviews= df_order_reviews[
    (df_order_reviews['review_comment_message'] != 'no message') &
    (df_order_reviews['review_comment_title'] != 'no title')].copy()

In [19]:
df_order_reviews

review_id                          order_id  \
9      8670d52e15e00043ae7de4c01cc2fe06  b9bf720beb4ab3728760088589c62129   
15     3948b09f7c818e2d86c9a546758b2335  e51478e7e277a83743b6f9991dbfa3fb   
19     373cbeecea8286a2b66c97b1b157ec46  583174fbe37d3d5f0d6661be3aad1786   
22     d21bbc789670eab777d27372ab9094cc  4fc44d78867142c627497b60a7e0228a   
34     c92cdd7dd544a01aa35137f901669cdf  37e7875cdce5a9e5b3a692971f370151   
...                                 ...                               ...   
99187  47e0954e156dac6512c25c6d2ecc1c66  16cbf959cfdb88c47ee2a29303547ec2   
99192  0e7bc73fde6782891898ea71443f9904  bd78f91afbb1ecbc6124974c5e813043   
99196  58be140ccdc12e8908ff7fd2ba5c7cb0  0ebf8e35b9807ee2d717922d5663ccdb   
99197  51de4e06a6b701cb2be47ea0e689437b  b7467ae483dbe956fe9acdf0b1e6e3f4   
99200  2ee221b28e5b6fceffac59487ed39348  f2d12dd37eaef72ed7b1186b2edefbcd   

       review_score       review_comment_title  \
9                 4                  recomendo   
15                5            Super recomendo   
19                1    Não chegou meu produto    
22                5                      Ótimo   
34                4                 Muito bom.   
...             ...                        ...   
99187             5               Nota máxima!   
99192             4                          👍   
99196             5         muito bom produto    
99197             3  Não foi entregue o pedido   
99200             2             Foto enganosa    

                                  review_comment_message  
9      aparelho eficiente. no site a marca do aparelh...  
15     Vendedor confiável, produto ok e entrega antes...  
19                                               Péssimo  
22                                          Loja nota 10  
34     Recebi exatamente o que esperava. As demais en...  
...                                                  ...  
99187  Muito obrigado,\r\n\r\nExcelente atendimento,b...  
99192                                          Aprovado!  
99196  Ficamos muito satisfeitos com o produto, atend...  
99197  Bom dia \r\nDas 6 unidades compradas só recebi...  
99200  Foto muito diferente principalmente a graninha...  

[9839 rows x 5 columns]

In [20]:
df_order_reviews.info()

<class 'pandas.DataFrame'>
Index: 9839 entries, 9 to 99200
Data columns (total 5 columns):
 #   Column                  Non-Null Count  Dtype
---  ------                  --------------  -----
 0   review_id               9839 non-null   str  
 1   order_id                9839 non-null   str  
 2   review_score            9839 non-null   int64
 3   review_comment_title    9839 non-null   str  
 4   review_comment_message  9839 non-null   str  
dtypes: int64(1), str(4)
memory usage: 461.2 KB


In [21]:
df_reviews=pd.read_csv("C:/Users/jwsam/Desktop/Ecom/kaggle_datasets/olist_order_reviews_dataset.csv")

In [22]:
df_reviews.drop(columns=['review_creation_date','review_answer_timestamp'],inplace=True)

In [23]:
df_reviews

review_id                          order_id  \
0      7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1      80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2      228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3      e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4      f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   
...                                 ...                               ...   
99219  574ed12dd733e5fa530cfd4bbf39d7c9  2a8c23fee101d4d5662fa670396eb8da   
99220  f3897127253a9592a73be9bdfdf4ed7a  22ec9f0669f784db00fa86d035cf8602   
99221  b3de70c89b1510c4cd3d0649fd302472  55d4004744368f5571d1f590031933e4   
99222  1adeb9d84d72fe4e337617733eb85149  7725825d039fc1f0ceb7635e3f7d9206   
99223  efe49f1d6f951dd88b51e6ccd4cc548f  90531360ecb1eec2a1fbb265a0db0508   

       review_score review_comment_title  \
0                 4                  NaN   
1                 5                  NaN   
2                 5                  NaN   
3                 5                  NaN   
4                 5                  NaN   
...             ...                  ...   
99219             5                  NaN   
99220             5                  NaN   
99221             5                  NaN   
99222             4                  NaN   
99223             1                  NaN   

                                  review_comment_message  
0                                                    NaN  
1                                                    NaN  
2                                                    NaN  
3                  Recebi bem antes do prazo estipulado.  
4      Parabéns lojas lannister adorei comprar pela I...  
...                                                  ...  
99219                                                NaN  
99220                                                NaN  
99221  Excelente mochila, entrega super rápida. Super...  
99222                                                NaN  
99223  meu produto chegou e ja tenho que devolver, po...  

[99224 rows x 5 columns]

In [24]:
df_reviews['review_comment_title'] = df_reviews['review_comment_title'].fillna("no title")

In [25]:
df_reviews

review_id                          order_id  \
0      7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1      80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2      228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3      e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4      f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   
...                                 ...                               ...   
99219  574ed12dd733e5fa530cfd4bbf39d7c9  2a8c23fee101d4d5662fa670396eb8da   
99220  f3897127253a9592a73be9bdfdf4ed7a  22ec9f0669f784db00fa86d035cf8602   
99221  b3de70c89b1510c4cd3d0649fd302472  55d4004744368f5571d1f590031933e4   
99222  1adeb9d84d72fe4e337617733eb85149  7725825d039fc1f0ceb7635e3f7d9206   
99223  efe49f1d6f951dd88b51e6ccd4cc548f  90531360ecb1eec2a1fbb265a0db0508   

       review_score review_comment_title  \
0                 4             no title   
1                 5             no title   
2                 5             no title   
3                 5             no title   
4                 5             no title   
...             ...                  ...   
99219             5             no title   
99220             5             no title   
99221             5             no title   
99222             4             no title   
99223             1             no title   

                                  review_comment_message  
0                                                    NaN  
1                                                    NaN  
2                                                    NaN  
3                  Recebi bem antes do prazo estipulado.  
4      Parabéns lojas lannister adorei comprar pela I...  
...                                                  ...  
99219                                                NaN  
99220                                                NaN  
99221  Excelente mochila, entrega super rápida. Super...  
99222                                                NaN  
99223  meu produto chegou e ja tenho que devolver, po...  

[99224 rows x 5 columns]

In [26]:
df_reviews['review_comment_message'] = df_reviews['review_comment_message'].fillna("no message")

In [27]:
df_reviews

review_id                          order_id  \
0      7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1      80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2      228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3      e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4      f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   
...                                 ...                               ...   
99219  574ed12dd733e5fa530cfd4bbf39d7c9  2a8c23fee101d4d5662fa670396eb8da   
99220  f3897127253a9592a73be9bdfdf4ed7a  22ec9f0669f784db00fa86d035cf8602   
99221  b3de70c89b1510c4cd3d0649fd302472  55d4004744368f5571d1f590031933e4   
99222  1adeb9d84d72fe4e337617733eb85149  7725825d039fc1f0ceb7635e3f7d9206   
99223  efe49f1d6f951dd88b51e6ccd4cc548f  90531360ecb1eec2a1fbb265a0db0508   

       review_score review_comment_title  \
0                 4             no title   
1                 5             no title   
2                 5             no title   
3                 5             no title   
4                 5             no title   
...             ...                  ...   
99219             5             no title   
99220             5             no title   
99221             5             no title   
99222             4             no title   
99223             1             no title   

                                  review_comment_message  
0                                             no message  
1                                             no message  
2                                             no message  
3                  Recebi bem antes do prazo estipulado.  
4      Parabéns lojas lannister adorei comprar pela I...  
...                                                  ...  
99219                                         no message  
99220                                         no message  
99221  Excelente mochila, entrega super rápida. Super...  
99222                                         no message  
99223  meu produto chegou e ja tenho que devolver, po...  

[99224 rows x 5 columns]

In [28]:
title_valid = (
    df_reviews['review_comment_title'].notna() &
    (df_reviews['review_comment_title'].str.strip().str.lower() != 'no title')
)
message_valid = (
    df_reviews['review_comment_message'].notna() &
    (df_reviews['review_comment_message'].str.strip().str.lower() != 'no message')
)
df_reviews['text_for_nlp'] = np.where(
    title_valid & message_valid,
    df_reviews['review_comment_title'].str.strip() + ". " +
    df_reviews['review_comment_message'].str.strip(),
    np.where(
        title_valid,
        df_reviews['review_comment_title'].str.strip(),
        np.where(
            message_valid,
            df_reviews['review_comment_message'].str.strip(),
            np.nan
        )
    )
)
df_nlp = df_reviews[df_reviews['text_for_nlp'].notna()].copy()

In [29]:
df_nlp.head(40)

review_id                          order_id  \
3   e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4   f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   
9   8670d52e15e00043ae7de4c01cc2fe06  b9bf720beb4ab3728760088589c62129   
12  4b49719c8a200003f700d3d986ea1a19  9d6f15f95d01e79bd1349cc208361f09   
15  3948b09f7c818e2d86c9a546758b2335  e51478e7e277a83743b6f9991dbfa3fb   
16  9314d6f9799f5bfba510cc7bcd468c01  0dacf04c5ad59fd5a0cc1faa07c34e39   
19  373cbeecea8286a2b66c97b1b157ec46  583174fbe37d3d5f0d6661be3aad1786   
22  d21bbc789670eab777d27372ab9094cc  4fc44d78867142c627497b60a7e0228a   
24  0e0190b9db53b689b285d3f3916f8441  79832b7cb59ac6f887088ffd686e1d5e   
27  fe3db7c069d694bab50cc43463f91608  2ca73e2ff9e3a186ad1e1ffb9b1d9c10   
28  e07eee7ad7666276bd8b27fbd83fd1ff  68b3cbc68e4cd1fe19bec35f1d3bb7ee   
29  2c5e27fc178bde7ac173c9c62c31b070  0ce9a24111d850192a933fcaab6fbad3   
32  58044bca115705a48fe0e00a21390c54  68e55ca79d04a79f20d4bfc0146f4b66   
34  c92cdd7dd544a01aa35137f901669cdf  37e7875cdce5a9e5b3a692971f370151   
36  08c9d79ec0eba1d252e3f52f14b8e6a9  e029f708df3cc108b3264558771605c6   
37  2170c9452c2216457b36003fe2969fbb  af32671d0b8ce510a94d9111d97371d3   
38  b193ff3c9f32a01f3a0d9ae26b94d244  e2e6ee1ed2d7f2f36b05d234983bd7a0   
39  9fd59cd04b42f600df9f25e54082a8d1  3c314f50bc654f3c4e317b055681dff9   
43  86c5cfa7fcbde303f704b60a78ced7d6  a6456e781cb962cc3f412b04de4fed7b   
47  890ec1767be8eb104877db6d68eaade5  b3271f5ebce80cbecb67c5b88487a44a   
49  7d9ac3e98adc0c8ad579e6a2257d79f9  dfa48852c93427261c98fb2e01f33bea   
50  c41bd21983e298a240410ebb391bf075  9562b424b05978a493db2a5122ecab1f   
51  e233e51d11511bf30e568c76360ace52  548df2c6e5f089574614894bca78acf5   
53  1edd1a85161468a77e5fcb165be9b294  d8e17cc85323061493b16b68a130bcff   
55  3d94fd645cdaacc8c9f0dc0a2a1f5166  4d483bf690ca21bdc005df9b623673c7   
59  500c05500aa275953129f49799ee5c73  8a9424899aac432d80d8e580932b5ee9   
61  6a0969d0769d2f3ff6438809a1cc455e  32638fbca8d1a197e649fd5138e12049   
62  8a64945e89526e68bc16383391590232  71960e6af960919256a958a3633e1040   
64  c0dbf9a325d2a2fece43873be47191e1  b324f4270850cbba23533b9a9d3434a9   
67  109b5ce2dd11bb8460eff3b86da6fefc  25362fbf6aac4b01a28dee1e076acc26   
68  6d06808638ec0701bccd70bc8d462c28  97d2f8fe76f2f253b8291e17b5383884   
69  750d4311e8ddb6cba549ee86f06f0cca  4b32a4cdfc1b80066a8ee3a4b7a2094f   
73  8b230a1373c6dc4bd867099fda1d7039  071251fe3b3493294536f03737a8a679   
75  cb2fc3e5711b5ae85e0491ee18af63ed  34e6d418f368f8079ae152bc178bc66a   
76  60c714ed14cef913944a3147094a4742  9ac05114800f02bfaa783bd76842dbe2   
79  c45811d9f90e22a81155b3a1e4a5c2e8  491f193fc52075598871cfe6f19976d4   
80  e29ab4a671400c41322e4855de633c6a  17b296b73ebefd6213299342b09f097f   
87  0bd4dcc4f6c4621baf37f73495cad8c4  a11cd01ac67beef7e8bf09740d8a35c1   
88  888f11883ff85fa5c07e6ae1192d85c4  248907ae8054cb8e93d5aa44857779b4   
89  65dfeb60c40e3cbb0a1838285d86f885  a2714ecbf6eeb3bb9cd7dba6dc1c5e82   

    review_score     review_comment_title  \
3              5                 no title   
4              5                 no title   
9              4                recomendo   
12             4                 no title   
15             5          Super recomendo   
16             2                 no title   
19             1  Não chegou meu produto    
22             5                    Ótimo   
24             5                 no title   
27             5                 no title   
28             5                 no title   
29             1                 no title   
32             1                 no title   
34             4               Muito bom.   
36             5                      Bom   
37             5                 no title   
38             5             Maravilhoso!   
39             1                 no title   
43             5         Entrega perfeita   
47             5                 no title   
49             5                 no title   
50

In [30]:
df_nlp.info()

<class 'pandas.DataFrame'>
Index: 42706 entries, 3 to 99223
Data columns (total 6 columns):
 #   Column                  Non-Null Count  Dtype
---  ------                  --------------  -----
 0   review_id               42706 non-null  str  
 1   order_id                42706 non-null  str  
 2   review_score            42706 non-null  int64
 3   review_comment_title    42706 non-null  str  
 4   review_comment_message  42706 non-null  str  
 5   text_for_nlp            42706 non-null  str  
dtypes: int64(1), str(5)
memory usage: 2.3 MB


## NLP and Sentiment Analysis
Using a multilingual transformer model to translate from Portugese to English then deduce the sentiment analysis score 

In [31]:
import torch

In [39]:
from transformers import MarianMTModel, MarianTokenizer
from tqdm import tqdm

In [ ]:
model_name = "Helsinki-NLP/opus-mt-ROMANCE-en"
token="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"
tokenizer = MarianTokenizer.from_pretrained(model_name,token=token)
model= MarianMTModel.from_pretrained(model_name, token=token)

In [ ]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
model.to(device)

### Test a Sample n=1000

In [ ]:
df_test = df_nlp.head(1000).copy()

def translate_batch(texts, batch_size=32):
    translated = []

    for i in tqdm(range(0, len(texts), batch_size)):
        batch = texts[i:i+batch_size]
        
        inputs = tokenizer(batch, return_tensors="pt", padding=True, truncation=True, max_length=512)
        inputs = {k: v.to(device) for k, v in inputs.items()}

        with torch.no_grad():
            outputs = model.generate(**inputs)

        decoded = tokenizer.batch_decode(outputs, skip_special_tokens=True)
        translated.extend(decoded)

    return translated

In [ ]:
import time
start = time.time()

_ = translate_batch(df_test["text_for_nlp"].tolist())

print("Time:", time.time() - start)